buildpack/files.go

```diff
@@ -198,3 +202,111 @@ type LayerMetadataFile struct {
 	Launch bool        `json:"launch" toml:"launch"`
 	Cache  bool        `json:"cache" toml:"cache"`
 }
+
+type typesTable struct {
+	Build  bool `toml:"build"`
+	Launch bool `toml:"launch"`
+	Cache  bool `toml:"cache"`
+}
+type layerMetadataTomlFile struct {
+	Data  interface{} `toml:"metadata"`
+	Types typesTable  `toml:"types"`
+}
+
+func (lmf *LayerMetadataFile) Encode(path, buildpackAPI string) error {
+	fh, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer fh.Close()
+	if supportsTypesTable(buildpackAPI) {
+		types := typesTable{Build: lmf.Build, Launch: lmf.Launch, Cache: lmf.Cache}
+		lmtf := layerMetadataTomlFile{Data: lmf.Data, Types: types}
+		return toml.NewEncoder(fh).Encode(lmtf)
+	}
+	return toml.NewEncoder(fh).Encode(lmf)
+}
+
+func DecodeLayerMetadataFile(path, buildpackAPI string) (LayerMetadataFile, error) {
+	fh, err := os.Open(path)
+	if os.IsNotExist(err) {
+		return LayerMetadataFile{}, nil
+	} else if err != nil {
+		return LayerMetadataFile{}, err
+	}
+	defer fh.Close()
+
+	if supportsTypesTable(buildpackAPI) {
+		var lmtf layerMetadataTomlFile
+		if _, err = toml.DecodeFile(path, &lmtf); err != nil {
+			return LayerMetadataFile{}, err
+		}
+		return LayerMetadataFile{Data: lmtf.Data, Build: lmtf.Types.Build, Launch: lmtf.Types.Launch, Cache: lmtf.Types.Cache}, nil
+	}
+	var lmf LayerMetadataFile
+	if _, err := toml.DecodeFile(path, &lmf); err != nil {
+		return LayerMetadataFile{}, err
+	}
+	return lmf, nil
+}
+
+func typesInTopLevel(path string) (bool, error) {
+	fh, err := os.Open(path)
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	defer fh.Close()
+
+	var lmf LayerMetadataFile
+	if _, err := toml.DecodeFile(path, &lmf); err != nil {
+		return false, err
+	}
+	return lmf.Cache || lmf.Launch || lmf.Build, nil
+}
+
+func typesInTypesTable(path string) (bool, error) {
+	fh, err := os.Open(path)
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	defer fh.Close()
+
+	var lmtf layerMetadataTomlFile
+	if _, err = toml.DecodeFile(path, &lmtf); err != nil {
+		return false, err
+	}
+	return lmtf.Types.Build || lmtf.Types.Launch || lmtf.Types.Cache, nil
+}
+
+func typesInRightFormat(path, buildpackAPI string) bool {
```

Comment:
I've made the separation between `typesInRightFormat` and `checkTypesFormat` because it felt more clear for me but I'm open to make changes.
Talking about your suggestion, I'm not sure how to actually do it while keeping it simple because for `buildpackAPI < 0.6`, we would like to send a warning if needed, and for `buildpackAPI >= 0.6`, we would like to return an error if needed.
What do you think?

Reply:
We could probably make the signature `validateLayerTypes(path, buildpackAPI string, out io.Writer
) error`. In the `< 0.6` it would never return an error but possibly write a warning.

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/537/commits/de2f8e688570dc8f2c17f94b92b944c5652cc6e0.